enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,133 @@
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes to add the support of RAID and BIOS configuration for baremetal IPI deployments.
+
+## Motivation
+
+With the new feature of [RAID configuration](https://github.com/metal3-io/baremetal-operator/pull/292) and the upcoming feature of [BIOS configuration](https://github.com/metal3-io/baremetal-operator/pull/302), openshift already has or will have the ability to configure RAID and BIOS for baremetal.
+
+However, these features only work with a running cluster, users cannot configure desired RAID and BIOS during IPI deployments.
+
+### Goals
+
+- Allow users to configure RAID and BIOS in ***install-config.yaml***.
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. Add two new fields *raid* and *firmware* to *platform.baremetal.hosts* in the [install-config.yaml](https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml) file.
+2. Openshift Installer process the field contents into manifests.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for master nodes and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Add two fields
+
+New *platform.baremetal.hosts* fields called *raid* and *firmware*.
+
+```yaml
+platform:
+  baremetal:
+    ...
+    hosts:
+      - name: master-0
+        role: master
+        bmc:
+          address: IPAddress
+          username: UserName
+          password: PassWord
+        bootMACAddress: MACAddress
+        raid: RAIDConfig
+        firmware: BIOSConfig
+```
+
+RAID feature has been implemented in **BMO**, so the *raid* field here is the same as the [BMH](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/config/crd/bases/metal3.io_baremetalhosts.yaml).
+The *firmware* field is the same as the *spec.firmware* field in **BMH** which is been advancing by [#302](https://github.com/metal3-io/baremetal-operator/pull/302).
+
+#### Process the fields in installer
+
+For master nodes, call the `BuildTargetRAIDCfg` method in **BMO** to process the *raid* field into *target_raid_config*, and finally write *target_raid_config* into the ***terraform.baremetal.auto.tfvars.json*** file.
```

Comment:
If we're going to import that code from the BMO into the installer, we will want to ensure that it is in a package with minimal dependencies (rather than in the controller implementation, for example). How is the code organized in the BMO today? Will we need to rearrange anything?

Reply:
> How is the code organized in the BMO today?

Now the raid related code is in the ironic package.

> Will we need to rearrange anything?

It may be better to create a new package for raid related code for reducing dependency.